File: java/zipper/src/main/java/Zipper.java
import java.util.Objects;

public class Zipper {

    private int value;
    public Zipper left;
    public Zipper right;
    public Zipper up;

    public Zipper(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
        this.up = null;
    }

    public int getValue() {
        return value;
    }

    public void setValue(int value) {
        this.value = value;
    }

    public void setLeft(Zipper left) {
        this.left = left;

        if (this.left != null) {
            this.left.setUp(this);
        }
    }

    public void setRight(Zipper right) {
        this.right = right;

        if (this.right != null) {
            this.right.setUp(this);
        }
    }

    public void setUp(Zipper up) {
        this.up = up;
    }

    public BinaryTree toTree() {

        Zipper zipper  = this.clone();

        while (zipper.up != null) {
            zipper = zipper.up;
        }

        return new BinaryTree(zipper);
    }

    public Zipper clone() {
        Zipper zipper = new Zipper(this.getValue());
        zipper.setLeft(this.left);
        zipper.setRight(this.right);
        zipper.setUp(this.up);

        return zipper;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Zipper zipper = (Zipper) o;
        return value == zipper.value && Objects.equals(left, zipper.left) && Objects.equals(right, zipper.right) && Objects.equals(up, zipper.up);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, left, right, up);
    }

    @Override
    public String toString() {
        return "value: " + this.value + ", " +
               "left: " + (left == null ? "null" : "{ " + left + " }") + ", " +
               "right: " + (right == null ? "null" : "{ " + right + " }");
    }
}

File: javascript/zipper/zipper.js
export class Zipper {

  constructor(tree, root, parent) {
    this.tree = tree;
    this.root = root || tree;
    this.parent = parent || null;
  }

  static fromTree(tree) {
    return new Zipper(JSON.parse(JSON.stringify(tree)));
  }

  toTree() {
    return this.root;
  }

  value() {
    return this.tree.value;
  }

  left() {
    return this.tree.left 
      ? new Zipper(this.tree.left, this.root, this.tree)
      : null;
  }

  right() {
    return this.tree.right
      ? new Zipper(this.tree.right, this.root, this.tree)
      : null;
  }

  up() {
    return this.parent
      ? new Zipper(this.parent, this.root)
      : null;
  }

  setValue(newValue) {
    this.tree.value = newValue;
    return new Zipper(this.tree, this.root, this.parent);
  }

  setLeft(newLeft) {
    this.tree.left = newLeft;
    return new Zipper(this.tree, this.root, this.parent);
  }

  setRight(newRight) {
    this.tree.right = newRight;
    return new Zipper(this.tree, this.root, this.parent);
  }

}
